### ISSUE TITLE
Identify and separate the _create_new_session into deferred calls

### ISSUE THREAD

**Issue by re-nobre** [pixelated/pixelated-user-agent#762]
## What
While working on the issue #759, @bwagnerr split the _get_leap_session call into two smaller ones, we now have setup_keymanager and _create_leap_session.

_create_leap_session is still taking more than our threshold of 100ms and we have to break it apart further. We run on @thaissiqueira machine using vagrant and we got 1.35724 seconds for this call. 

## Definition of done
The goal for this issue is:
- Try to split the _create_new_session into deferred calls. 
- If these new deferred calls still take more than 100ms. Try to split it again. 
- If it's not possible to split, separate them out from the main processing thread, like the following: 
   - if it's a processing intense call, than we should defer to a process using _spawnProcess_
   - if it's a disk IO, than we should defer to a new thread using deferToThread
   - if it's a HTTP request, than we should use twisted.web.client

**Comment by olabini** [comment]
I disagree with this definition - the goal here is NOT to decrease the time. The goal is to decrease the time this function and its parts blocks the main reactor thread. This is NOT a performance goal, it is a scalability goal.

**Comment by bwagnerr** [comment]
@olabini I tried to rewrite this to make it clearer that we are talking strictly about timing of execution on the main thread, does it make sense to you?

Title
---
Break down the _create_leap_session call running on the main thread

What
---
We refactored the get_leap_session to defer the blocking calls it does and now it takes less than 100ms to run.
After this change, we have two former pieces of it still taking more than 100ms, they are:

**setup_keymanager** and **_create_leap_session**

Each one of these run on the main thread for more than the 100ms threshold we are aiming for and should be split further.

Definition of done
---
The _create_leap_session call on login takes less than our threshold of 100ms on the main thread

**Comment by olabini** [comment]
I'm not sure I like this much better. It's not very clear and doesn't get to the core of the problem. If you take something that takes 2s to run, and split that into 30 pieces, that is still a problem. The real task here is to identify the blocking components of these calls, and separate them out from the main processing thread so that it doesn't end up blocking it. The goal isn't to just make it smaller, the goal is to make it smaller in such a way that we drastically reduce the amount of blocking code.

**Comment by olabini** [comment]
I think the previous description was completely fine, except for the last sentence of "The time should be < 0.1 seconds". Also, remember, stories are placeholders for conversation. They don't have to be exhaustive, they should just not be wrong - and I added the initial comment because I wanted to make sure everyone understood the real goal here.

**Comment by mnandri** [comment]
pushed the defer_to_process of scrypt.hash in this branch https://github.com/pixelated/pixelated-user-agent/tree/defer_to_process
missing  measurements/benchmarking... and probably refactoring

**Comment by re-nobre** [comment]
hey @olabini! 

Thank you for your comments. It was a good conversation on the issue and we also had more conversation here. 

I will rename this issue and edit the definition of done. 
Please, tell me if the new one makes more sense:

**New name: **
Identify and separate the _create_new_session into deferred calls

**New definition of done:**
The goal for this issue is:

- Try to split the _create_new_session into deferred calls. 
- If these new deferred calls still take more than 100ms. Try to split it again. 
- If it's not possible to split, separate them out from the main processing thread, like the following: 
   - if it's a processing intense call, than we should defer to a process using _spawnProcess_
   - if it's a disk IO, than we should defer to a new thread using deferToThread
   - if it's a HTTP request, than we should use twisted.web.client 

Does it make sense for you?

Cheers,

**Comment by olabini** [comment]
Hi,

That all makes a lot of sense to me. The only thing I'm wondering is about disk IO - I'm not sure deferToThread is the right Twisted solution, but that's something the team will have investigate.

Thanks.

**Comment by re-nobre** [comment]
Thank you, @olabini ! :)

**Comment by tuliocasagrande** [comment]
As @mnandri said in https://github.com/pixelated/pixelated-user-agent/issues/763#issuecomment-242169959, we identified a blocking GET request inside `self._provider.smtp_info()`.

https://github.com/pixelated/pixelated-user-agent/commit/debb86dfc9cf70d310bd3b45dc32c91d5d26b9d1 moves this request to the UA initialization.

**Comment by thaissiqueira** [comment]
We got a new patch (#768) to more accurately measure inlineCallbacks.

@mnandri Can you try this patch and see what else is taking more than 100ms? If not, could you lower the threshold a bit and find the next longest call?

**Comment by mnandri** [comment]
With the newer patch,
I obtain, single user login on dev - sorted by timing, above 5ms:
```
0.00629 EXECUTING inlineCallback: <code object _key_exists at 0x7fba50d03b30, file "pixelated/bitmask_libraries/keymanager.py", line 43> in thread 140438341924608
0.00645 EXECUTING inlineCallback: <code object generate_openpgp_key at 0x7fba50d03ab0, file "pixelated/bitmask_libraries/keymanager.py", line 35> in thread 140438341924608
0.00658 EXECUTING inlineCallback: <code object finish_bootstrap at 0x7fba50cf9930, file "pixelated/config/sessions.py", line 152> in thread 140438341924608
0.00720 EXECUTING deferred: <bound method DeferredList._cbDeferred of <DeferredList at 0x7fba4c273950>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:882) in thread 140438341924608
0.00744 EXECUTING deferred: <function _emit_done_data_sync at 0x7fba4c3b7d70> (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/api.py:768) in thread 140438341924608
0.00842 EXECUTING deferred: <bound method Deferred.callback of <Deferred at 0x7fba4c29a8c0 waiting on Deferred at 0x7fba4c29aea8>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:397) in thread 14043834190.00849 EXECUTING deferred: <bound method DeferredList._cbDeferred of <DeferredList at 0x7fba4c37aef0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:882) in thread 140438341924608
0.00985 EXECUTING deferred: <bound method OpenPGPScheme._migrate_documents_schema of <leap.keymanager.openpgp.OpenPGPScheme object at 0x7fba4c3c20d0>> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/openpgp.py:110) in thread
0.01096 EXECUTING deferred: <bound method DeferredList._cbDeferred of <DeferredList at 0x7fba4c3e9830>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:882) in thread 140438341924608
0.01231 EXECUTING deferred: <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7fba535675d0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in thre
0.01971 EXECUTING deferred: <function execute at 0x7fba4c3049b0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1353) in thread 140438341924608
0.02067 EXECUTING inlineCallback: <code object _put at 0x7fba518b16b0, file "/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/__init__.py", line 265> in thread 140438341924608
0.02073 EXECUTING deferred: <function send at 0x7fba4c3b7d70> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/__init__.py:352) in thread 140438341924608
0.02172 EXECUTING deferred: <function execute at 0x7fba4c38c5f0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1353) in thread 140438341924608
0.02258 EXECUTING inlineCallback: <code object get_sync_info at 0x7fba51b70c30, file "/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py", line 86> in thread 140438341924608
0.02267 EXECUTING inlineCallback: <code object sync at 0x7fba51b9b0b0, file "/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py", line 59> in thread 140438341924608
0.02281 EXECUTING inlineCallback: <code object sync at 0x7fba5238d7b0, file "/root/junk/git2016-08-26 14:16:14 [leap.soledad.client.sync] DEBUG Soledad target sync info:
0.02299 EXECUTING deferred: <function execute at 0x7fba4c38ca28> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1353) in thread 140438341924608
0.02818 EXECUTING deferred: <function execute at 0x7fba4c3e8488> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1353) in thread 140438341924608
0.02917 EXECUTING inlineCallback: <code object get_sync_info at 0x7fba51b70c30, file "/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py", line 86> in thread 140438341924608
0.02926 EXECUTING inlineCallback: <code object sync at 0x7fba51b9b0b0, file "/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py", line 59> in thread 140438341924608
0.03002 EXECUTING inlineCallback: <code object sync at 0x7fba5238d7b0, file "/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py", line 482> in thread 140438341924608
0.03022 EXECUTING deferred: <function execute at 0x7fba4c3e85f0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1353) in thread 140438341924608
0.03152 EXECUTING inlineCallback: <code object first_required_sync at 0x7fba50cf98b0, file "pixelated/config/sessions.py", line 147> in thread 140438341924608
0.22494 EXECUTING deferred: <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7fba53567590>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in thre
0.28997 EXECUTING deferred: <bound method DeferredList._cbDeferred of <DeferredList at 0x7fba4c305cb0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:882) in thread 140438341924608
```

**Comment by tuliocasagrande** [comment]
We used the following steps previously described at https://github.com/pixelated/pixelated-user-agent/issues/759#issuecomment-241900981:

- Pre-create the users using the [create_users.py](https://github.com/pixelated/pixelated-user-agent/blob/master/service/test/reactor/create_users.py) script
- Add the metric of the login time on [login_resource.py](https://github.com/pixelated/pixelated-user-agent/blob/master/service/pixelated/resources/login_resource.py#L151)
```python
    @defer.inlineCallbacks
    def _handle_login(self, request):
        self.creds = self._get_creds_from(request)
        t = Clock('login')
        iface, leap_session, logout = yield self._portal.login(self.creds, None, IResource)
        t.stop()
        defer.returnValue(leap_session)
```

- Run the [concurrent_logins.rb](https://github.com/pixelated/pixelated-user-agent/blob/master/service/test/reactor/concurrent_logins.rb) to login 1, 2, 5 and 10 users before and after https://github.com/pixelated/pixelated-user-agent/commit/debb86dfc9cf70d310bd3b45dc32c91d5d26b9d1
- Lastly, collect the measurements on the ~/MetricsTime file.

We ran that inside a vagrant box using dev.pixelated-project.org as the provider and we restarted the UA before each measurement.

----

We collected the following results:

| 1 user   | Before           | After            |
|--------|--------------|-------------|
|1st user|6.77930 secs |5.49107 secs|

| 2 users | Before     | After             |
|---------|------------|------------|
|1st user |8.37305 secs|5.23895 secs|
|2nd user |7.60592 secs|5.78893 secs|

| 5 users | Before        | After               |
|---------|---------------|--------------|
|1st user |11.53506 secs  |5.38747 secs  |
|2nd user |10.75737 secs  |5.33523 secs  |
|3rd user |10.33379 secs  |3.86871 secs  |
|4th user |9.03913  secs  |4.99343 secs  |
|5th user |7.81097  secs  |3.43645 secs  |

|10 users  | Before        | After               |
|----------|---------------|--------------|
|1st user  |15.85027 secs  |5.50054 secs  |
|2nd user  |13.93355 secs  |3.55027 secs  |
|3rd user  |16.54212 secs  |4.71445 secs  |
|4th user  |12.49138 secs  |5.94653 secs  |
|5th user  |10.40710 secs  |3.89094 secs  |
|6th user  |13.49211 secs  |4.26614 secs  |
|7th user  |16.56385 secs  |3.87987 secs  |
|8th user  |12.22430 secs  |4.33805 secs  |
|9th user  |7.66550  secs  |4.02319 secs  |
|10th user |11.48437 secs  |5.05553 secs  |

**Comment by mnandri** [comment]
Run on dev,
with a single user login, timing measurements (above 5ms) shows:
```
0.0071 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7f835019b098>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140202167281408
0.0074 haha callback: function <bound method Deferred.callback of <Deferred at 0x7f83500e8a28 waiting on Deferred at 0x7f83500e8fc8>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:399) in thread 1402021
0.0078 haha callback: function <function _emit_done_data_sync at 0x7f83501c00c8> (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/api.py:768) in thread 140202167281408
0.0084 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7f83501b0ab8>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140202167281408
0.0091 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7f83500c1680>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140202167281408
0.0104 haha callback: function <bound method OpenPGPScheme._migrate_documents_schema of <leap.keymanager.openpgp.OpenPGPScheme object at 0x7f8350207f10>> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/openpgp.py:110) in thr
0.0118 haha callback: function <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7f83563b65d0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in t
0.0203 haha callback: function <function execute at 0x7f83501761b8> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140202167281408
0.0208 haha callback: function <function execute at 0x7f83501db500> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140202167281408
0.0212 haha inlineCallback: function _put (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/__init__.py:265) in thread 140202167281408
0.0213 haha callback: function <function send at 0x7f83501c00c8> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/__init__.py:352) in thread 140202167281408
0.0215 haha inlineCallback: function get_sync_info (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py:86) in thread 140202167281408
0.0215 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py:59) in thread 140202167281408
0.0216 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140202167281408
0.0218 haha callback: function <function execute at 0x7f83501db7d0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140202167281408
0.0331 haha callback: function <function execute at 0x7f83502375f0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140202167281408
0.0338 haha inlineCallback: function get_sync_info (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py:86) in thread 140202167281408
0.0416 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py:59) in thread 140202167281408
0.0430 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140202167281408
0.0433 haha callback: function <function execute at 0x7f83501db5f0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140202167281408
0.0443 haha inlineCallback: function first_required_sync (pixelated/config/sessions.py:147) in thread 140202167281408
0.0444 haha inlineCallback: function create_leap_session (pixelated/config/leap.py:41) in thread 140202167281408
0.2348 haha callback: function <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7f83563b6590>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in t
0.3024 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7f83500e8ea8>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140202167281408
```
Of that, we deduce, what else could be investigated, by order of priority:
* whoever calls `DeferredList._cbDeferred` most likely `defer.gatherResults` which is called by `LeapMailstore.get_mails`
* `first_required_sync` which is called by `create_leap_session`
* `get_sync_info` which is called by Soledad/SqlCipher.sync
* `execute` callback -- no clue
* `_put` which is called by `send` callback in `Keymanager.send_key`

**Comment by mnandri** [comment]
Same experiment, with 10 users (hatch rate: 1user/sec)
```
0.0300 haha callback: function <function execute at 0x7fc3fc0e8230> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0303 haha inlineCallback: function create_leap_session (pixelated/config/leap.py:41) in thread 140479959545600
0.0307 haha inlineCallback: function first_required_sync (pixelated/config/sessions.py:147) in thread 140479959545600
0.0308 haha inlineCallback: function create_leap_session (pixelated/config/leap.py:41) in thread 140479959545600
0.0309 haha callback: function <function execute at 0x7fc3dc77d500> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0312 haha callback: function <function execute at 0x7fc3fc47aed8> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0312 haha callback: function <function execute at 0x7fc3fc55f9b0> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0312 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140479959545600
0.0315 haha callback: function <function execute at 0x7fc3fc0e8758> (/root/junk/git_env/local/lib/2016-08-27 19:17:43 [leap.soledad.client.pragmas] DEBUG SQLCIPHER: SETTING WRITE-AHEAD LOGGING
0.0321 haha inlineCallback: function get_sync_info (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py:86) in thread 140479959545600
0.0322 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py:59) in thread 140479959545600
0.0323 haha inlineCallback: function get_sync_info (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py:86) in thread 140479959545600
0.0323 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140479959545600
0.0324 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py:59) in thread 140479959545600
0.0325 haha callback: function <function execute at 0x7fc3fc0dc410> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0325 haha inlineCallback: function get_sync_info (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py:86) in thread 140479959545600
0.0326 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py:59) in thread 140479959545600
0.0329 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140479959545600
0.0331 haha callback: function <function execute at 0x7fc3dc7c6938> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0333 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140479959545600
0.0335 haha callback: function <function execute at 0x7fc3fc58e578> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0343 haha inlineCallback: function first_required_sync (pixelated/config/sessions.py:147) in thread 140479959545600
0.0344 haha inlineCallback: function create_leap_session (pixelated/config/leap.py:41) in thread 140479959545600
0.0389 haha callback: function <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7fc403f0f5d0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in t
0.0410 haha inlineCallback: function first_required_sync (pixelated/config/sessions.py:147) in thread 140479959545600
0.0411 haha inlineCallback: function create_leap_session (pixelated/config/leap.py:41) in thread 140479959545600
0.0448 haha callback: function <function execute at 0x7fc3dc749e60> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0458 haha inlineCallback: function get_sync_info (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/http_target/api.py:86) in thread 140479959545600
0.0459 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sync.py:59) in thread 140479959545600
0.0467 haha inlineCallback: function sync (/root/junk/git_env/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py:482) in thread 140479959545600
0.0469 haha callback: function <function execute at 0x7fc3fc162de8> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1355) in thread 140479959545600
0.0480 haha inlineCallback: function first_required_sync (pixelated/config/sessions.py:147) in thread 140479959545600
0.0481 haha inlineCallback: function create_leap_session (pixelated/config/leap.py:41) in thread 140479959545600
0.2099 haha callback: function <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7fc403f0f590>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in t
0.2792 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3dc6fccb0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.2850 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3fc0fc638>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.2950 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3fc042ea8>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.2956 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3fc404ef0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.2981 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3dc779440>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.3215 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3dc7b0cf8>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.3301 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3fc052440>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.3372 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3fc0a0e18>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
0.3551 haha callback: function <bound method DeferredList._cbDeferred of <DeferredList at 0x7fc3dc779488>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:884) in thread 140479959545600
1 haha callback: function <bound method KeyDocumentsMigrator._migrate_docs of <leap.keymanager.migrator.KeyDocumentsMigrator object at 0x7fc3fc3ad310>> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/migrator.py:68) in thr
```
which confirms the same list as above... but aggravated, in terms of magnitude.

**Comment by olabini** [comment]
Wait wait wait wait here. @tuliocasagrande  - the code sample you posted have this:
`t = Clock('login')
 iface, leap_session, logout = yield self._portal.login(self.creds, None, IResource)
 t.stop()`

This makes no sense to me - it won't actually measure the time of login, it will measure how long the reactor thread is spending outside of the _handle_login function while the yield is happening. What is the goal here?

**Comment by tuliocasagrande** [comment]
Hello @olabini, I didn't understand your point.

Our goal with the measurement was to get the overall login time indeed, but we consider that the `yield` is suspending the execution of the `_handle_login` and waiting until `self._portal.login(self.creds, None, IResource)` finishes to stop the clock.

Can you clarify if you understood differently?

**Comment by olabini** [comment]
Hi,

Maybe I misunderstood. I thought you wanted to know exactly how long self._portal.login takes. In that case you are still measuring the wrong thing, since the clock.stop will only happen once login() has finished AND everything else on the reactor thread in the meantime has finished.

But more importantly, from other issues it seems that this is used to measure complete login time. As I think I mentioned in another issue, I feel like this should be done using the complete http request time, to avoid missing other places with contention.

**Comment by bwagnerr** [comment]
@olabini I like the idea of looking at the http request time, but keeping in mind that doesn't cover the time the user waits at the loading page.

**Comment by olabini** [comment]
? what does that mean?

**Comment by bwagnerr** [comment]
@olabini the http request finishes the moment the loading page is shown, not when the user arrives at the inbox.

**Comment by olabini** [comment]
Yes, but for testing for contention during the login process, that should be fine.

**Comment by bwagnerr** [comment]
We finished this part of the work and got the contention problem controlled with few users (1,2,5 users). We are now testing with more users on issue #771 to see what is the current behaviour and are accounting for these concerns about how to test the whole login process on those tests.

@olabini
